cashaddr.md

```diff
@@ -0,0 +1,117 @@
+
+The payload is composed of 3 elements:
+1. A version byte indicating the type of address.
+2. A hash.
+3. A 40 bits checksum.
+
+#### Version byte
+
+The version byte's most signficant bit is reserved and must be 0. The 4 next bits indicate the type of address and the 3 least significant bits indicate the size of the hash.
+
+| Size bits | Hash size in bits |
+| --------: | ----------------: |
+|         0 |               160 |
+|         1 |               192 |
+|         2 |               224 |
+|         3 |               256 |
+|         4 |               320 |
+|         5 |               384 |
+|         6 |               448 |
+|         7 |               512 |
+
+Encoding the size of the hash in the version field ensure that it is possible to check that the length of the address is correct.
+
+| Type bits |      Meaning      | Version byte value |
+| --------: | :---------------: | -----------------: |
+|         0 |       P2KH        |                  0 |
+|         1 |       P2SH        |                  8 |
+
+Further types will be added as news features are added.
+
+#### Hash
+
+The hash part really deserve not much explaination as its meaning is dependent on the version field. It is the hash that represents the data, namely a hash of the public key for P2KH and the hash of the reedemScript for P2SH.
+
+#### Checksum
+
+The checksum is a 40 bits BCH codes defined over GF(2^5). It ensure the detection of up to 6 errors in the address and 8 in a row. Combined with the length check, this provides very strong guarantee against errors.
+
+The checksum is computed per the code bellow:
+
+````cpp
+uint64_t PolyMod(const data &v) {
+    uint64_t c = 1;
+    for (uint8_t d : v) {
+        uint8_t c0 = c >> 35;
+        c = ((c & 0x07ffffffff) << 5) ^ d;
+        
+        if (c & 0x01) c ^= 0x98f2bc8e61;
+        if (c & 0x01) c ^= 0x79b76d99e2;
+        if (c & 0x01) c ^= 0xf33e5fb3c4;
+        if (c & 0x01) c ^= 0xae2eabe2a8;
+        if (c & 0x01) c ^= 0x1e4f43e470;
+    }
+    
+    return c ^ 1;
+}
+````
+
+The data to be fed are:
+1. The lower 5 bits of each characters of the prefix.
+2. A zero for the separator.
+3. The payload by chunks of 5 bits.
+
+## Error correction
+
+BCH codes allows for error correction. However, it is strongly advised that error correction is not done in an automatic manner as it may cause funds to be lost irrecoverably if done incorrectly. It may however be used to hint a user at a possible error.
+
+## Uppercase/lowercase
+
+Lower case is preferred for cashaddr, but uppercase is accepted. A mixture of lower case and uppercase must be rejected.
+
+Allowing for uppercase ensure that the address can be encoded effisciently in QR codes using the alphanumeric mode[[2]](#alphanumqr).
+
+## Double prefix
+
+In some contexts, such as payment URLs or QR codes, the addresses are currently prefixed with `bitcoincash:`. In these contexts, the address must not be double prefixed.
+
+## References
```

Comment:
As indicated, a good reference for BCH codes linked to introduction would be nice.

Reply:
Addressed in recent update.